### ISSUE TITLE
Provide ConcurrentDictionary the ability to set initial capacity with DefaultConcurrencyLevel

### ISSUE THREAD

**Issue by dotnetchris** [dotnet/corefx#1218]
It's absurdly difficult to create a ConcurrentDictionary with an initial capacity. This forces you to set the arcane DefaultConcurrencyLevel integer. I absolutely want to defer that to the platform. This leads to using this horrific code:
```c#
(int)typeof(ConcurrentDictionary<,>).MakeGenericType(typeof(int), typeof(int))
    .GetProperty("DefaultConcurrencyLevel", BindingFlags.Static | BindingFlags.NonPublic)
    .GetValue(new ConcurrentDictionary<int, int>());
```

So either add a constructor that allows this. Add magic number support that providing < 1 results in this, or just make DefaultConcurrencyLevel be a public property.

**Comment by joshfree** [comment]
Adding a ctor overload to specify the initial capacity seems pretty reasonable.  @dotnetchris Do you want to take a crack at the API proposal (https://aka.ms/apireview) and expand upon your post?

**Comment by stephentoub** [comment]
Part of the difficulty is we've had requests for the opposite, e.g. specifying a concurrencyLevel but not a capacity.  Given they're both the same data type, there could be confusion if a ctor was added for just one of them, and as concurrencyLevel comes first in the existing signature, it would be the most likely to assume in that position for a ctor with just one Int32 parameter.  There are of course other design options, but this is part of why it hasn't yet been added previously.
cc: @pharring

**Comment by dotnetchris** [comment]
It sounds like just making `DefaultConcurrencyLevel` public would be the simplest.

**Comment by Priya91** [comment]
@dotnetchris Do you have a proposal for this?

**Comment by dotnetchris** [comment]
@Priya91 what does that mean?

**Comment by Priya91** [comment]
@dotnetchris Can you come up with an api-review speclet that describes the problem, and the proposed API changes? Like [here](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md)

**Comment by dotnetchris** [comment]
@Priya91 is the short of it "fork Corefx, make a local change to 'fix it'" then comment back here?

**Comment by Priya91** [comment]
> is the short of it "fork Corefx, make a local change to 'fix it'" then comment back here?

Did you get a chance to look at the api proprosal doc that was linked in the previous comment? From the conversation I gather your proposal to fix the issue is to expose ```DefaultConcurrencyLevel```, so that can be used in the (int,int) ctor, instead of hardcoded user value for concurrencylevel. So you can update your issue description with the api speclet, like this,

```c#
namespace System.Collections.Concurrent
{
    public class ConcurrentDictionary<TKey, TValue> : IDictionary<TKey, TValue>, IDictionary, IReadOnlyDictionary<TKey, TValue>
   {
       public static int DefaultConcurrentLevel { get { return default(int); } }
       public static int DefaultCapacity { get { return default(int); } } // For the other case.
   }
}
```

And when the issue is marked with api-ready-for-review tag, the api review board will review the addition.

**Comment by dotnetchris** [comment]
@Priya91 yes i reviewed the link you provided and the response you provided does not match what i took away from that link at all.

Re-reading that for a 3rd time, i see you're now referencing a link in link the "here's a good example". That's very non-obvious and something much better needs integrated into the post directly. Or atleast make it scream CLICK HERE FOR AN EXAMPLE.

**Comment by Priya91** [comment]
@dotnetchris Thanks for the feedback.

cc @terrajobst @weshaggard: Can you take a look at documentation feedback on Api-Review process. Thanks!

**Comment by safern** [comment]
@dotnetchris are you still interested in this one? If so, could you please update the issue description to have a formal API proposal?
